forge-game/src/main/java/forge/game/ability/effects/AnimateEffectBase.java

```diff
@@ -79,15 +84,36 @@ public static void doAnimate(final Card c, final SpellAbility sa, final Integer
         }
 
         if (!addType.isEmpty() || !removeType.isEmpty() || addAllCreatureTypes || !remove.isEmpty()) {
+            if (perpetual) {
+                Map <String, Object> params = new HashMap<>();
+                params.put("AddTypes", addType);
+                params.put("RemoveTypes", removeType);
+                params.put("RemoveXTypes", remove);
+                params.put("Timestamp", timestamp);
+                c.addPerpetual(Pair.of("Types", params));
```

Comment:
for readability there's no point in wrapping them in pairs on all these calls
doing it in `addPerpetual` instead would be even easier